### ISSUE TITLE
ETher thief : please also check for unprotected withdraw balances.

### ISSUE THREAD

**Issue by ytrezq** [ConsenSys/mythril#1371]
Currently, `ether_thief` only checks for balances whose amount can be specified directly. It doesn’t check for balances in storage which aren’t restricted against growing indefinitely.

One can imagine such fair ponzi scheme :

    library SafeMath {

        function mul(uint256 a, uint256 b)
        internal
        pure
        returns(uint256 c) {
            if(a == 0) {
                    return 0;
            }
            c = a * b;
            require(c / a == b, "mul failed");
            return c;
        }
        
        function sub(uint256 a, uint256 b)
        internal
        pure
        returns(uint256 c) {
            require(b <= a, "sub failed");
            c = a - b;
            require(c <= a, "sub failed");
            return c;
        }

        function add(uint256 a, uint256 b)
        internal
        pure
        returns(uint256 c) {
            c = a + b;
            require(c >= a, "add failed");
            return c;
        }

    }

    contract Ponzi {
        using SafeMath for * ;
        struct purchase_data {
            uint256 time;
            uint256 value;
        }

        uint256 public PERIOD = 79488000; // 3 months
        mapping (address => purchase_data[]) public history;
        uint256 public MAXIMUM = 1000000000000000000;
        uint256 public MINIMUM = 100000000000000000;
        mapping (address => uint) public balanceOf;

        function play() public payable {
            require(msg.value == (msg.value/100000000000000000).mul(100000000000000000), "use a number with no more than 2 decimals");
            require(msg.value <= MAXIMUM, "max is 1");
            require(msg.value >= MINIMUM, "min is 0.1");
            history[msg.sender].push(Player_data(now,msg.value));
        }

        function Earn_interests() public {
            require(history[msg.sender][_index].time.add(PERIOD) >= now, "You can’t get your investment yet.");
            uint256 index = history[msg.sender].length.sub(1);
            require(history[msg.sender][_index].is_withdraw == false, "already withrawn")

            balanceOf[msg.sender] = balanceOf[msg.sender].add(history[msg.sender][_index].value.add(history[msg.sender][_index].value.mul(2)/100)); // 2% interests every 3 months
            
            // history[msg.sender][_index].value = 0;
        }

        function withdraw() public {
            uint256 index = history[msg.sender].length.sub(1);
            history[msg.sender][_index].value = 0;

            msg.sender.transfer(balanceOf[msg.sender]);
            balanceOf[msg.sender] = 0;
        }
}

Because `// history[msg.sender][_index].value = 0;` is eronously commented out, one can call `Earn_interests()` several successive times (as much as he/she wants to withraw) which enable to take all the money owned by the contract (making possibly a 10000% return instead of the normal 2%).

And mythril isn’t detecting that specific error even with ` -b 8 -t 5`.

**Comment by norhh** [comment]
EtherThief should be able to handle such cases now, the cause is most probably solver timeout, constraints with multiplication is expensive, we are currently looking on how to speed it up, meanwhile, you can try increasing solver timeout with `--solver-timeout <time in ms>`

**Comment by ytrezq** [comment]
@norhh except it doesn’t. Did you try without? Also, I don’t have enough memory to do this without state merge. And your state merge branch requires the infura update to work. I tried to backport the patch but it doesn’t work.

**Comment by norhh** [comment]
`Did you try without? ` solver timeout defaults to 10s if it isn't mentioned, so if you want to run you should use the `--solver-timeout` flag

**Comment by norhh** [comment]
Cases where the vulnerability requires more than 5 transactions aren't supported by Mythril at the moment. MythX should support it in the coming months with symbolic summaries.

**Comment by ytrezq** [comment]
> Cases where the vulnerability requires more than 5 transactions take a long time for Mythril at the moment. MythX might handle them well in the coming months.

@norhh as you can see in the source code above, with 4 transactions you steal more than what you have (just a tiny %).

In theory, the number of transactions in order to steal everything can be infinite if we suppose the number of Ethereums on the contract is infinite.

**Comment by norhh** [comment]
`as you can see in the source code above, with 4 transactions you steal more than what you should have (just a tiny %).` yeah. At the moment it takes forever to run mythril for 4 transactions due to multiplication and division, we are working on speeding it up.

**Comment by ytrezq** [comment]
@norhh not that much. Please backport the latest changes to https://github.com/ConsenSys/mythril/pull/1281 in order to be able to use that branch again (it no longer works because of the lastest Infura changes).

**Comment by ytrezq** [comment]
@norhh just a question, the z3 solver can be run in multithreaded mode (not Mythril itself bu z3) https://stackoverflow.com/a/53246261. Why don’t you set up the correct parameters in order to gain performances for such scenarios?

**Comment by norhh** [comment]
@ytrezq , There was a reasonable slowdown when I tried it on Mythril. So z3's multithreading isn't that useful at the moment.

**Comment by ytrezq** [comment]
@norhh even on 32 core cpu? I mean can the number of cpus overcome this slowdown at some point?

**Comment by norhh** [comment]
I am not sure on a 32 core pc, but my pc has 16 cores and that was the case. Anyway, I will add a command-line flag for it which can switch on this feature.

**Comment by ytrezq** [comment]
@norhh hyperthreaded? If yes, 32 threads or 8 cores with 16 threads Please notice there are many scenarios where hyperthreading decreases performance. Especially with all those recent security patches.

What’s possible then is each thread adds something like 45% overhead while each pair of threads with intel’s hyperthreading increase performances at 30% maximum compared with hyperthreading disabled.

**Comment by norhh** [comment]
yes the timeouts were increased and I think it's hyperthreaded. Anyway, the latest develop branch should have the parallelisation option `--parallel-solving`

**Comment by ytrezq** [comment]
@norhh both OsX Windows bds and Linux have the possibility for unprivileged software to set its cpu affinity (bind threads to specific threads). There are many C software negatively affected by hyperthreading which will automatically restrict their number of threads and bind them to only physical cpu numbers.

Until this option is added, one should be to cap the maximum number of threads z3 is using through the Mythril’s command line.

**Comment by ytrezq** [comment]
@norhh since those parameters can t be tuned on mythx and that mythril can t go fast enough because it analyse duplicate sates, this should definitely explicitely added.
